apiserver: graceful termination with new option shutdown-send-retry-after

tkashem opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Introduce a new server run option `shutdown-send-retry-after` 
```
If true the HTTP Server will continue listening until all non longrunning request(s) in flight 
have been drained, during this window all incoming requests will be rejected with
a status code 429 and a 'Retry-After' response header.
```
The default value is set to `false` to maintain current behavior.

With this option, we now have two modes of graceful termination:
- `shutdown-send-retry-after=true`: we initiate shutdown of the HTTP Server  when all non `longrunning` requests in-flight have been drained.
- `shutdown-send-retry-after=false`: we initiate shutdown of the HTTP Server  as soon as `shutdown-delay-duration` has elapsed. (This is the current behavior)

The following happens with `shutdown-send-retry-after=true`:
- Do not call http `Server.Shutdown` immediately after  `ShutdownDelayDuration` elapses. Rather, wait for all non `longrunning` requests to drain. (the requests that are being counted by `WithWaitGroup` filter)
- Immediately after `ShutdownDelayDuration` elapses, any new incoming request  should be answered with a `429` and the following headers: 
- `Connection: close`  and 
- `Retry-After: N`

#### Which issue(s) this PR fixes:

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
a new server run option 'shutdown-send-retry-after'  has been introduced. If true the HTTP Server
will continue listening until all non longrunning request(s) in flight have been drained, during this window all 
incoming requests will be rejected with a status code 429 and a 'Retry-After' response header.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

tkashem commented:
/test all

tkashem commented:
/assign @smarterclayton

smarterclayton commented:
Hrm, I'd prefer 429 - try again (every request after shutdown we reject is " come back later) and connection close ensures that.

fedebongio commented:
/triage accepted

tkashem commented:
> Hrm, I'd prefer 429 - try again (every request after shutdown we reject is " come back later) and connection close ensures that.

@smarterclayton  (this was a typo in my PR description) yes, we return `429`

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -233,6 +233,18 @@ type Config struct {
+	// then initiate a shutdown of the HTTP Server.
+	// If false, we initiate a shutdown of the HTTP Server
+	// as soon as ShutdownDelayDuration has elapsed.
+	DelayServerShutdown bool
```
Can we invert the logic in the var name and the flag? `keepListeningDuringGracefulTermination`.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -233,6 +233,18 @@ type Config struct {
+
+	// GracefulShutdown provides access to the various shutdown events
+	// that happen during the graceful termination of the apiserver.
+	GracefulShutdown GracefulShutdown
```
is this nil before shutdown?

sttts replied:
wondering that this thing is and how to name it better. It is not "a shutdown".

tkashem replied:
it is not `nil`, we create a new instance of it inside `config.Complete`. we need it instanced up front so we can use the events in handler chain build function.

tkashem replied:
> wondering what this thing is and how to name it better. It is not "a shutdown".

`GracefulShutdownCoordinator`, `GracefulShutdownDriver`? let me know if you have any suggestions :)

tkashem replied:
addressed in the refactor pr

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go:
```diff
@@ -0,0 +1,57 @@
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithNotAcceptingNewRequest(handler http.Handler, notAcceptingNewRequest <-chan struct{}) http.Handler {
```
<bike-shedding>With429After</bike-shedding> ?

tkashem replied:
How about `WithRetryAfter`?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go:
```diff
@@ -0,0 +1,57 @@
+		var reject bool
+		select {
+		case <-notAcceptingNewRequest:
+			reject = true
```
```
handler.ServeHTTP(w, req)
return
```

p0lyn0mial commented:
I am not sure if I understand why we need an additional option.

IIRC the maximum time a request can take is `60` seconds (`RequestTimeout`) (excluding long-running requests). So a request accepted at `T+70` will have been terminated at `T+70+60`.

What we could do is to:

- set the `ShutdownTimeout` to `70` seconds to give more time to the server to send a response for very late requests `(70+60)`

- send `429` immediately after starting the termination process to reduce the potential number of in-flight requests to handle.

I haven't seen code that actually determines the number of requests in-flight but it looks like a broken LB or a client that uses an API directly can significantly delay or even block termination of an API server. That seems very powerful.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/graceful_shutdown.go:
```diff
@@ -0,0 +1,307 @@
+
+	// Events returns the ShutdownEvents associated. The caller can
+	// can use to get notified when the shutdown events take place.
+	Events() ShutdownEvents
```
do we really need Events() and the return value of Start for the same thing? Looks like these two overlap.

tkashem commented:
@sttts I am going to break it down to multiple PRs so it is easy to review - the refactor should be a separate PR and then a follow up PR for `keep-listening-during-graceful-termination`

tkashem commented:
/assign @sttts @p0lyn0mial

tkashem commented:
no flakes so far:
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationEnabled
...
23m0s: 7240 runs so far, 0 failures
23m5s: 7268 runs so far, 0 failures
23m10s: 7296 runs so far, 0 failures
23m15s: 7320 runs so far, 0 failures
23m20s: 7348 runs so far, 0 failures
```

tkashem commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -351,7 +359,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+		//   requests in flight have been drained
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 5 * time.Second
```
it changes the previous behavior, it is not backward compatible, not sure it has to be
- it changes the maximum time from `60s` to `65s`
- doesn't apply the timeout set in `ShutdownTimeout` (could be set to 10s)

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -785,6 +795,9 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
+	if c.KeepListeningDuringGracefulTermination {
+		handler = genericfilters.WithRetryAfter(handler, c.terminationSignals.AfterShutdownDelayDuration.Signaled())
```
please add a comment that this handler must be **before** the `WithWaitGroup` otherwise the server can deadlock

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,52 @@
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
```
This filter changes the status codes previously returned on paths like `livez` and `readyz` previously `500` now it will be `429`. Again this might be not a backward compatible change/behavriour.

p0lyn0mial replied:
We might actually allow some actors to get through like `the loopback client`. Otherwise, the server might not shut down gracefully.

sttts replied:
readyz/livez/healthyz/metrics definitely should not change

sttts replied:
+1 for loopback client

p0lyn0mial replied:
I updated the filter to check the path and the user, PTAL https://github.com/kubernetes/kubernetes/pull/103357

sttts replied:
the `shutdownDelayDurationElapsedCh` is a detail of the caller, not the func. Would call the func `WithRetryAfterDuringShutdown(handler, shutdownCh)`. Then it is understandable without context.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+	if s.KeepListeningDuringGracefulTermination {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing
+		//   requests in flight have been drained
```
be clear what this means for long-running requests

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing
+		//   requests in flight have been drained
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
```
what do the 5s mean? Why not immediately or 1s?

tkashem replied:
I set it to `2s`, net/http waits for `1s` after sending GO_AWAY to take into account round trip time for the peer, so we should respect that.

sttts replied:
please add this in the comment. 5s seems to be arbitrary without explanation. Also sync comment and code :)

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -614,6 +622,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 		StorageVersionManager: c.StorageVersionManager,
 
 		Version: c.Version,
+
```
nit: we have many empty lines in here that do not make sense :)

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,142 @@
+// construct the Retry-After response.
+type retryAfterConditionFunc func() (*retryAfterParams, bool)
+
+func (f retryAfterConditionFunc) ShouldSendRetryAfterResponse() (*retryAfterParams, bool) {
```
what does "after" mean here? Is this `RespondWithRetryAfter`?

tkashem replied:
renamed it to `ShouldRespondWithRetryAfter`

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,142 @@
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, func(r *http.Request) bool {
+		if hasExcludedPathPrefix(r) {
```
`return hasExcludedPathPrefix || isKubeApiserverUserAgent`

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,142 @@
+func withRetryAfter(handler http.Handler, excludeFn excludeFunc, retryAfterFn retryAfterConditionFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := retryAfterFn.ShouldSendRetryAfterResponse()
+		if !sendRetryAfter {
```
if !sendRetryAfter || excludeFn(req)

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 2 * time.Second
+		klog.V(1).InfoS("[graceful-termination] HTTP Server", "ShutdownTimeout", shutdownTimeout)
```
The `[graceful-termination]` prefix is used everywhere?

tkashem replied:
yes, so we can search with it during debugging

sttts replied:
I mean the syntax. It looks special. `Using graceful termination http server shutdown timeout` would be more normal.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -74,18 +74,18 @@ func (w *wrappedTerminationSignal) Signal() {
-func wrapTerminationSignals(t *testing.T, ts *lifecycleSignals, callback func(bool, string, lifecycleSignal)) {
-	newWrappedTerminationSignal := func(delegated lifecycleSignal) lifecycleSignal {
-		return &wrappedTerminationSignal{
+func wrapLifecycleSignals(t *testing.T, ts *lifecycleSignals, callback func(bool, string, lifecycleSignal)) {
```
factor these renames out into their own commit

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,142 @@
+		// if we are here this means it's time to send Retry-After response to
+		// all requests except the ones that are excluded.
+		if excludeFn(req) {
+			handler.ServeHTTP(w, req)
```
I think that it will keep the `drained` queue busy and will block the server from shutting down, am I right?

sttts replied:
we can exlucde them from the wait group too

tkashem replied:
yeah, for now using the same exclude function in `WithWaitGroup`, I have left a todo to merge these two filters in the future.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:
```diff
@@ -245,5 +255,9 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"will return success, but /readyz immediately returns failure. Graceful termination starts after this delay "+
 		"has elapsed. This can be used to allow load balancer to stop sending traffic to this server.")
 
+	fs.BoolVar(&s.KeepListeningDuringGracefulTermination, "keep-listening-during-shutdown", s.KeepListeningDuringGracefulTermination, ""+
```
sync with @p0lyn0mial about the names. They should look symmetric.

tkashem committed:
apiserver: rename test variables

tkashem committed:
apiserver: add a new mode for graceful termination

add a new mode for graceful termination with the new server run option
'shutdown-send-retry-after'
- shutdown-send-retry-after=true: we initiate shutdown of the
  HTTP Server when all in-flight request(s) have been drained. during
  this window all incoming requests are rejected with status code
  429 and the following response headers:
    - 'Retry-After: N' - client should retry after N seconds
    - 'Connection: close' - tear down the TCP connection
- shutdown-send-retry-after=false: we initiate shutdown of the
  HTTP Server as soon as shutdown-delay-duration has elapsed. This
  is in keeping with the current behavior.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go:
```diff
@@ -41,21 +47,32 @@ func WithWaitGroup(handler http.Handler, longRunning apirequest.LongRunningReque
+			// shutdown delay duration has elapsed and SafeWaitGroup.Wait has been invoked
+			// we are going to allow execution of the requests that WithRetryAfter is
+			// excluding from sending a Retry-After response.
+			if excludeFn(req) {
```
why is this not like long-running? Can we combine both predicates?

i.e. a function called excludeFunc in withWaitGroup and then it is not really excluded.

tkashem replied:
we don't want to combine with `longRunning` - `excludeFn` is relevant when shutdown delay duration has elapsed. Only then `WithRetryAfter` starts responding with retry-after response. 
inside `withWaitGroup` we are not sending retry-after for the white-listed (allowed) requests and we are not counting them either.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,141 @@
+var (
+	// health probes and metrics scraping are never rejected, we will continue
+	// serving these requests after shutdown delay duration elapses.
+	excludedPathPrefixes = []string{
```
excluded from what? This variable is public for the whole filters package. It needs a more specialized name.

tkashem replied:
renamed

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,141 @@
+
+func (f retryAfterConditionFunc) ShouldRespondWithRetryAfter() (*retryAfterParams, bool) {
+	return f()
+}
```
odd construct. Just use `retryAfterConditionFunc()` directly. No need for a method.

tkashem replied:
done

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,141 @@
+
+func withRetryAfter(handler http.Handler, excludeFn excludeFunc, retryAfterFn retryAfterConditionFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := retryAfterFn.ShouldRespondWithRetryAfter()
```
retryAfterFn()

tkashem commented:
/retest

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -352,7 +361,22 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+		//   net/http waits for 1s for the peer to respond to a GO_AWAY frame, so
+		//   we should wait for a minimum of 2s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 2 * time.Second
```
the long-running connections should have a chance to terminate gracefully.
with `shutdownTimeout = 2 * time.Second` this is impossible.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,140 @@
+// NOTE: both WithRetryAfter and WithWaitGroup must use the same exact isRequestExemptFunc 'isRequestExemptFromRetryAfter,
+// otherwise SafeWaitGroup might wait indefinitely and will prevent the server from shutting down gracefully.
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, isRequestExemptFromRetryAfter, newRetryAfterFunc(shutdownDelayDurationElapsedCh))
```
nit: move the isRequestExemptFromRetryAfter comment above this line, out of the func doc

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,140 @@
+
+func withRetryAfter(handler http.Handler, isRequestExemptFn isRequestExemptFunc, shouldRespondWithRetryAfterFn shouldRespondWithRetryAfterFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := shouldRespondWithRetryAfterFn()
```
`params, send := ...`. The func is about retryAfter. No need for fully qualified var names.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,140 @@
+// newRetryAfterFunc builds a shouldRespondWithRetryAfterFunc that starts
+// sending Retry-After response to all incoming requests once the shut down
+// delay duration (ShutdownDelayDuration) has elapsed.
+func newRetryAfterFunc(shutdownDelayDurationElapsedCh <-chan struct{}) shouldRespondWithRetryAfterFunc {
```
haven't checked where this is used outside of WithRetryAfter. Just inline this there. This abstraction does not help, but rather distributes complexity.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:
```diff
@@ -0,0 +1,140 @@
+// NOTE: both 'WithRetryAfter' and 'WithWaitGroup' filters should use this function
+// to exempt the set of requests from being rejected or tracked.
+func isRequestExemptFromRetryAfter(r *http.Request) bool {
+	if isKubeApiserverUserAgent(r) || hasExemptPathPrefix(r) {
```
return isKubeApiserverUserAgent(r) || hasExemptPathPrefix(r)

sttts commented:
/lgtm
/approve
